src/main/kotlin/com/pubnub/api/retry/RetryableCallback.kt

```diff
@@ -0,0 +1,88 @@
+package com.pubnub.api.retry
+
+import org.slf4j.LoggerFactory
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import kotlin.random.Random
+
+internal abstract class RetryableCallback<T>(
+    retryPolicy: RequestRetryPolicy,
+    endpointGroupName: RetryableEndpointGroup,
+    private val call: Call<T>,
+    private val isEndpointRetryable: Boolean
+) : Callback<T>, RetryableBase<T>(retryPolicy, endpointGroupName) {
+    private val log = LoggerFactory.getLogger(this.javaClass.simpleName)
+    private var retryCount = 0
+    private val maxRetryNumber = getRetryCount()
+    private val random = Random.Default
+    private var exponentialMultiplier = 0.0
+
+    override fun onResponse(call: Call<T>, response: Response<T>) {
+        if (shouldRetryOnResponse(response)) {
+            retryOnResponseWithError(response)
+        } else {
+            onFinalResponse(call, response)
+        }
+    }
+
+    override fun onFailure(call: Call<T>, t: Throwable) {
+        if (shouldRetryOnFailure(t)) {
+            retryOnFailure()
+        } else {
+            onFinalFailure(call, t)
+        }
+    }
+
+    private fun shouldRetryOnResponse(response: Response<T>): Boolean {
+        return !response.isSuccessful &&
+            retryCount < maxRetryNumber &&
+            isErrorCodeRetryable(response.raw().code) &&
+            isRetryPolicySetForThisRestCall() &&
+            isEndpointRetryable
+    }
+
+    private fun shouldRetryOnFailure(t: Throwable): Boolean {
+        val exception = Exception(t)
+        return retryCount < maxRetryNumber &&
+            isExceptionRetryable(exception) &&
+            isRetryPolicySetForThisRestCall() &&
+            isEndpointRetryable
+    }
+
+    private fun isExceptionRetryable(e: Exception): Boolean {
+        return e.cause?.let { cause ->
+            retryableExceptions.any { it.isInstance(cause) }
+        } ?: false
+    }
+
+    private fun retryOnFailure() {
+        val effectiveDelay = getDelayInMilliSecForRetryOnFailure()
+        retry(effectiveDelay)
+    }
+
+    private fun retryOnResponseWithError(response: Response<T>) {
+        val effectiveDelay = getDelayInMilliSecForRetryOnResponse(response)
+        retry(effectiveDelay)
+    }
+
+    private fun retry(delayInMilliSec: Int) {
+        retryCount++
+        val randomDelayInMilliSec = random.nextInt(BOUND)
+        val effectiveDelay = delayInMilliSec + randomDelayInMilliSec
+        log.trace("Added random delay so effective retry delay is $effectiveDelay")
+        Thread.sleep(effectiveDelay.toLong())
```

Comment:
You can't sleep here (or do any other blocking operations), as this runs on the Main thread on Android and will freeze the UI (and lead to Application Not Responding)
See: https://square.github.io/retrofit/2.x/retrofit/index.html?retrofit2/Callback.html

You either need to find a different way or change the executor when creating the Retrofit instance. Although, by now people might be depending in their apps on the fact that all PubNub callbacks run on the Main thread, and changing that might be problematic, we should probably leave it and rewrite the retry code.

Reply:
Good catch. I will add:
        executorService.execute {
            try {
                Thread.sleep(effectiveDelay.toLong())
                call.clone().enqueue(this)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }